### PR TITLE
On-board Compass Calibration

### PR DESCRIPTION
**Pull request by bugobliterator**
Following are the series of patches to do onboard compass calibration as in 3DR Solo. The implementation details can be found in CompassCalibrator.cpp. Features and how to wire compass cal are as follows:
* The compass calibrator does calculation of both hard and soft iron coefficients
* proper spread of samples has been ensured (read comments in CompassCalibrator.cpp for explanation). The distribution progress is sent out as a part of MAG_CAL_PROGRESS mavlink message, checkout ardupilotmega.xml for other fields and their explanation.
* after the completion of calibration subsequent MAG_CAL_REPORT is sent, please refer ardupilotmega.xml for details and explanation of each field in the message
* For starting calibration MAV_CMD_DO_START_MAG_CAL is sent from GCS, the command consists of options: calibration mask, auto retry, autosave, delay, auto reboot. similarly to accept and cancel mag cal MAV_CMD_DO_ACCEPT_MAG_CAL and MAV_CMD_DO_CANCEL_MAG_CAL can be supplemented from GCS.

Note: The compass calibrator has only been tested for pixhawk and pixhawk-v2 board especially on Solo and has been found to work seamlessly. Currently there is no GCS which supports this calibration process. GCS developers please contact @bugobliterator or @jschall if any assistance is required for the implementation.

### CODE REVIEW

**Comment by tridge** [comment]
Thanks Sid!
I've started the review. I'll add comments on the patches

File: ArduCopter/compass_cal.cpp [hunk 1]
```diff
@@ -0,0 +1,18 @@
+/// -*- tab-width: 4; Mode: C++; c-basic-offset: 4; indent-tabs-mode: nil -*-
+
+#include "Copter.h"
+
+#define COMPASS_CAL_STICK_DELAY 2.0f // 2 seconds
+#define COMPASS_CAL_DELAY 5.0f
+
+void Copter::compass_cal_update() {
+    compass.compass_cal_update();
+    static bool cal_has_run = false;
```
Comment (tridge) [inline ArduCopter/compass_cal.cpp]: why is this state stored here? Can't the compass library do the reboot itself if the user has requested it?

File: ArduCopter/motors.cpp [hunk 2]
```diff
@@ -727,6 +727,20 @@ bool Copter::arm_checks(bool display_failure, bool arming_from_gcs)
...
+    if(compass.is_calibrating()) {
+        if (display_failure) {
+            gcs_send_text_P(MAV_SEVERITY_CRITICAL,PSTR("Arm: Compass calibration running"));
+        }
+        return false;
+    }
+
+    if(compass.compass_cal_requires_reboot()) {
+        if (display_failure) {
+            gcs_send_text_P(MAV_SEVERITY_CRITICAL,PSTR("Arm: Compass calibrated requires reboot"));
```
Comment (tridge) [inline ArduCopter/motors.cpp]: thanks for adding this!

File: libraries/AP_Compass/AP_Compass_Calibration.cpp [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+        if (_calibrator[i].running()) {
+            AP_Notify::flags.compass_cal_running = 1;
+        }
+    }
+}
+
+bool
+Compass::start_calibration(uint8_t i, bool retry, bool autosave, float delay, bool autoreboot)
+{
+    if (healthy(i)) {
```
Comment (tridge) [inline libraries/AP_Compass/AP_Compass_Calibration.cpp]: I would prefer if (!healthy(i)) { return false; } as it makes for easier to read code (less indentation)

Comment (bugobliterator) [inline libraries/AP_Compass/AP_Compass_Calibration.cpp]: agreed

File: libraries/AP_Compass/AP_Compass_Calibration.cpp [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+            cal_status == COMPASS_CAL_RUNNING_STEP_ONE ||
+            cal_status == COMPASS_CAL_RUNNING_STEP_TWO) {
+            uint8_t completion_pct = cal.get_completion_percent();
+            uint8_t completion_mask[10];
+            Vector3f direction(0.0f,0.0f,0.0f);
+            uint8_t attempt = cal.get_attempt();
+
+            memset(completion_mask, 0, sizeof(completion_mask));
+
+            mavlink_msg_mag_cal_progress_send(
```
Comment (tridge) [inline libraries/AP_Compass/AP_Compass_Calibration.cpp]: why do we have the completion_mask and direction when they are always zero in this code?

File: libraries/AP_Compass/AP_Compass_Calibration.cpp [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+
+void
+Compass::send_mag_cal_progress(mavlink_channel_t chan)
+{
+    uint8_t cal_mask = get_cal_mask();
+
+    for (uint8_t i=0; i<COMPASS_MAX_INSTANCES; i++) {
+        CompassCalibrator& cal = _calibrator[i];
+
+        uint8_t& compass_id = i;
```
Comment (tridge) [inline libraries/AP_Compass/AP_Compass_Calibration.cpp]: why is this a reference?

File: libraries/AP_Compass/AP_Compass_AK8963.cpp [hunk 6]
```diff
@@ -211,8 +211,7 @@ void AP_Compass_AK8963::read()
 #if CONFIG_HAL_BOARD_SUBTYPE == HAL_BOARD_SUBTYPE_LINUX_BEBOP
     field.rotate(ROTATION_YAW_90);
 #endif
-
-    publish_field(field, _compass_instance);
+    publish_filtered_field(field, _compass_instance);
```
Comment (tridge) [inline libraries/AP_Compass/AP_Compass_AK8963.cpp]: don't we need rotate_field() and correct_field() here?

**Comment by tridge** [comment]
i've done the initial review. I think it looks really good apart from the issues I marked in the individual code comments

**Comment by bugobliterator** [comment]
Great!! I'll fix those issue within this week and ping you back for the merge.

File: libraries/AP_Compass/AP_Compass_Calibration.cpp (hunk 4 above)
Comment (bugobliterator) [inline libraries/AP_Compass/AP_Compass_Calibration.cpp]: currently they are not implemented but might be implemented in the future if required. completion_percent covers the information to user for now.

File: libraries/AP_Compass/AP_Compass_Calibration.cpp (hunk 5 above)
Comment (bugobliterator) [inline libraries/AP_Compass/AP_Compass_Calibration.cpp]: there is no reason for them to be there, I'll remove them

File: libraries/AP_Compass/AP_Compass_AK8963.cpp (hunk 6 above)
Comment (bugobliterator) [inline libraries/AP_Compass/AP_Compass_AK8963.cpp]: those operations are being done during publication of unfiltered/raw fields

Comment (bugobliterator) [inline libraries/AP_Compass/AP_Compass_AK8963.cpp]: although that is being done twice for HMC so I'll correct that

File: ArduCopter/compass_cal.cpp (hunk 1 above)
Comment (bugobliterator) [inline ArduCopter/compass_cal.cpp]: I didn't believe compass library to be doing such tasks and that rebooting and other higher tasks should be done at vehicle end.

**Comment by tridge** [comment]
Thanks for all your work on this Sid, I've pushed to master

**Pull request closed without merging**